primitives/src/vrf.rs

```diff
@@ -0,0 +1,100 @@
+use num::rational::BigRational;
+use num::ToPrimitive;
+use parity_scale_codec::{Decode, Encode};
+use sp_arithmetic::per_things::{PerThing, Perbill, Perquintill};
+use sp_runtime_interface::pass_by::PassByCodec;
+use sp_runtime_interface::runtime_interface;
+
+pub mod sortition {
+	use super::*;
+	/// S is a hyperparameter representing participation rate.
+	/// The higher the value, the higher the chances of being sampled.
+	/// R is the prover's relative stake. the output is proportional to the stake.
+	pub fn model(s: Perquintill, r: Perquintill) -> f64 {
+		use pdf::probability_density_function;
+		probability_density_function(s.into(), r.into()).into()
+	}
+
+	#[derive(Encode, Decode, PassByCodec)]
+	pub struct Wrap(Perquintill);
+
+	impl From<Wrap> for f64 {
+		fn from(w: Wrap) -> Self {
+			w.0.to_sub_1_float()
+		}
+	}
+
+	impl From<Perquintill> for Wrap {
+		fn from(p: Perquintill) -> Self {
+			Wrap(p)
+		}
+	}
+
+	#[runtime_interface]
+	pub trait Pdf {
+		fn probability_density_function(sample: Wrap, weight: Wrap) -> Wrap {
+			let complement = 1f64 - sample.0.to_sub_1_float();
+			let r: f64 = weight.into();
+			let f: f64 = complement.powf(r);
+			Perquintill::from_float(1f64 - f).into()
+		}
+	}
+
+	pub fn threshold(sample_size: Perquintill, stake: u128, total_stake: u128) -> u128 {
+		let ratio = Perquintill::from_rational(stake, total_stake);
+		let level = model(sample_size, ratio);
+		let level = BigRational::from_float(level).expect("Model codomain [0,1)");
+		let threshold = u128::MAX / level.denom() * level.numer();
+		threshold.to_u128().expect("n * X, X ~ [0,1), n: u128; qed.")
+	}
+
+	#[cfg(test)]
+	mod tests {
+		use super::*;
+
+		#[test]
+		fn distribution_works() {
```

Comment:
I suggest a better name to make it clear which is the subject under test. Something like

```suggestion
		fn sortition_model_should_return_expected_distribution() {
```